rm-datasource/src/main/java/io/seata/rm/datasource/xa/ResourceManagerXA.java

```diff
@@ -0,0 +1,93 @@
+ *
+ * @author sharajava
+ */
+public class ResourceManagerXA extends AbstractDataSourceCacheResourceManager {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ResourceManagerXA.class);
+
+    @Override
+    public void init() {
+        LOGGER.info("ResourceManagerXA init ...");
+
+    }
+
+    @Override
+    public BranchType getBranchType() {
+        return BranchType.XA;
+    }
+
+    @Override
+    public BranchStatus branchCommit(BranchType branchType, String xid, long branchId, String resourceId,
+                                     String applicationData) throws TransactionException {
+        return finishBranch(true, branchType, xid, branchId, resourceId, applicationData);
+    }
+
+    @Override
+    public BranchStatus branchRollback(BranchType branchType, String xid, long branchId, String resourceId,
+                                       String applicationData) throws TransactionException {
+        return finishBranch(false, branchType, xid, branchId, resourceId, applicationData);
+    }
+
+    private BranchStatus finishBranch(boolean committed, BranchType branchType, String xid, long branchId, String resourceId,
+                                       String applicationData) throws TransactionException {
+        XAXid xaBranchXid = XAXidBuilder.build(xid, branchId);
+        Resource resource = dataSourceCache.get(resourceId);
+        if (resource instanceof AbstractDataSourceProxyXA) {
+            try (ConnectionProxyXA connectionProxyXA = ((AbstractDataSourceProxyXA)resource).getConnectionForXAFinish(xaBranchXid)) {
+                if (committed) {
+                    connectionProxyXA.xaCommit(xid, branchId, applicationData);
+                    LOGGER.info(xaBranchXid + " was committed.");
+                    return BranchStatus.PhaseTwo_Committed;
+                } else {
+                    connectionProxyXA.xaRollback(xid, branchId, applicationData);
+                    LOGGER.info(xaBranchXid + " was rolled back.");
+                    return BranchStatus.PhaseTwo_Rollbacked;
+                }
+            } catch (XAException xe) {
+                throw new TransactionException("XA Exception " + xe.getMessage(), xe);
+            } catch (SQLException sqle) {
+                if (committed) {
+                    LOGGER.info(xaBranchXid + " commit failed since " + sqle.getMessage(), sqle);
+                    // FIXME: case of PhaseTwo_CommitFailed_Unretryable
+                    return BranchStatus.PhaseTwo_CommitFailed_Retryable;
+                } else {
+                    LOGGER.info(xaBranchXid + " rollback failed since " + sqle.getMessage(), sqle);
+                    // FIXME: case of PhaseTwo_RollbackFailed_Unretryable
+                    return BranchStatus.PhaseTwo_RollbackFailed_Retryable;
+                }
+            }
+        } else {
+            throw new TransactionException("Unknown Resource for XA resource " + resourceId + " " + resource);
```

Comment:
throw exception will result in encode error.

Reply:
In fact, this is a case that SHOULD NEVER happen.

Just merged XAException and  SQLException into retryable status.